ddsrecorder_participants/src/cpp/mcap/McapHandler.cpp

```diff
@@ -0,0 +1,460 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file McapHandler.cpp
+ */
+
+#define MCAP_IMPLEMENTATION  // Define this in exactly one .cpp file
+
+#include <cpp_utils/exception/InitializationException.hpp>
+#include <cpp_utils/exception/InconsistencyException.hpp>
+
+#include <fastrtps/types/DynamicType.h>
+
+#include <ddspipe_core/types/dynamic_types/schema.hpp>
+
+#include <ddsrecorder_participants/mcap/McapHandler.hpp>
+
+namespace eprosima {
+namespace ddsrecorder {
+namespace participants {
+
+using namespace eprosima::ddspipe::core::types;
+
+McapHandler::McapHandler(
+        const char* file_name,
+        std::shared_ptr<ddspipe::core::PayloadPool> payload_pool,
+        unsigned int max_pending_samples,
+        unsigned int buffer_size,
+        unsigned int downsampling,
+        unsigned int event_window)
+        // bool autostart /* = false */)
+    : payload_pool_(payload_pool)
+    , max_pending_samples_(max_pending_samples)
+    , buffer_size_(buffer_size)
+    , downsampling_(downsampling)
+    , event_window_(event_window)
+{
+    auto status = mcap_writer_.open(file_name, mcap::McapWriterOptions("ros2"));
+    if (!status.ok()) {
+        throw utils::InitializationException(
+            STR_ENTRY << "Failed to open MCAP file " << file_name << " for writing: " << status.message);
+    }
+
+    logInfo(DDSRECORDER_MCAP_HANDLER,
+        "MCAP file <" << file_name << "> .");
+}
+
+McapHandler::~McapHandler()
+{
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Destroying handler.");
+
+    if (paused_)
+    {
+        stop_event_thread_();
+    }
+    else
+    {
+        dump_data_();
+    }
+    mcap_writer_.close();
+}
+
+void McapHandler::add_schema(const fastrtps::types::DynamicType_ptr& dynamic_type)
+{
+    assert(nullptr != dynamic_type);
+    std::string type_name = dynamic_type->get_name();
+    {
+        // Check if it exists already
+        // NOTE: must be unique mutex taken because it could write afterwards
+        std::unique_lock<SchemaMapType> lock(schemas_);
+        if (schemas_.find(type_name) != schemas_.end())
+        {
+            return;
+        }
+
+        // Schema not found, generate from dynamic type and store
+        std::string schema_text = generate_ros2_schema(dynamic_type);
+
+        // TODO remove
+        logInfo(DDSRECORDER_MCAP_HANDLER, "\nAdding schema with name " << type_name << " :\n" << schema_text << "\n");
+
+        // Create schema and add it to writer and to schemas map
+        mcap::Schema new_schema(type_name, "ros2msg", schema_text);
+
+        {
+            std::lock_guard<std::mutex> guard(write_mtx_);
+            mcap_writer_.addSchema(new_schema);
+        }
+
+        schemas_.insert({type_name, std::move(new_schema)});
```

Comment:
Be careful about moving this. I do not know why but `addSchema` gets a non const reference.
Add a comment about it at least